### PR TITLE
Users routes

### PR DESCRIPTION
**Pull request by Layvier**

### CODE REVIEW

File: src/api/users/auth.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        err => new HashingError()
+    )
+
+export const hashPassword = (salt: string) => (pwd: string) =>
+    TE.tryCatch(
+        () => hash(pwd, salt),
+        err => new HashingError()
+    )
+
+export const encryptPassword = (pwd: string) => TE.chain((salt: string) => hashPassword(salt)(pwd))(generateSalt())
```
Comment (L-oris) [inline src/api/users/auth.ts]: https://ramdajs.com/docs/#curry

File: src/api/registerRoute.ts [hunk 2]
```diff
@@ -46,4 +86,10 @@ const registerRouter = (app: Koa, router: Router) => app.use(router.routes())
 const registerKoaRoute = ({ app }: Context) => (routesConfig: RouteConfig[]) =>
     registerRouter(app, addRoutesToRouter(createRouter(), routesConfig))
 
+export const registerRoute = <A extends t.Any, B extends t.Any, C extends t.Any, D extends t.Any>(
+    config: RouteConfig<A, B, C, D>
```
Comment (L-oris) [inline src/api/registerRoute.ts]: I don't see any point of using `A` `B` `C` `D` if they're all typed as `any` and they don't have meaningful names

File: src/api/registerRoute.ts [hunk 3]
```diff
@@ -12,19 +14,57 @@ export declare type Response<B extends object> = {
...
-        const resp = await pipe(handler(context), catchError)()
+const koaContextToRequest = (ctx: Router.RouterContext): Request => {
+    return {
+        body: ctx.request['body'] as object,
+        query: ctx.request.query,
+        params: ctx.params,
+    }
+}
+
+class ValidationError extends Error {
```
Comment (L-oris) [inline src/api/registerRoute.ts]: there's `errors.ts` for errors

File: src/api/registerRoute.ts [hunk 4]
```diff
@@ -12,19 +14,57 @@ export declare type Response<B extends object> = {
...
+    constructor(m) {
```
Comment (L-oris) [inline src/api/registerRoute.ts]: what does `m` stand for? I'm curious :)

File: src/api/registerRoute.ts [hunk 5]
```diff
@@ -12,19 +14,57 @@ export declare type Response<B extends object> = {
...
+    Params extends object = {}
+> = (request: Request<RequestBody, Query, Params>) => TE.TaskEither<AnyError, Response<ResponseBody>>
 
-interface RouteConfig<ResponseBody extends object = {}> {
+export interface RouteConfig<
+    ResponseBody extends t.Any = t.Any,
+    RequestBody extends t.Any = t.Any,
+    Query extends t.Any = t.Any,
+    Params extends t.Any = t.Any
+> {
```
Comment (L-oris) [inline src/api/registerRoute.ts]: are we sure that that we need all these generics? Which value are they providing?

File: src/domain/services/auth.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        err => new HashingError()
+    )
+
+export const hashPassword = (salt: string) => (pwd: string) =>
+    TE.tryCatch(
+        () => hash(pwd, salt),
+        err => new HashingError()
+    )
+
+export const encryptPassword = (pwd: string) => TE.chain((salt: string) => hashPassword(salt)(pwd))(generateSalt())
```
Comment (L-oris) [inline src/domain/services/auth.ts]: https://ramdajs.com/docs/#curry ?

File: src/api/registerRoute.ts (hunk 4 above)
Comment (Layvier) [reply src/api/registerRoute.ts]: message ^^

File: src/api/registerRoute.ts (hunk 5 above)
Comment (Layvier) [reply src/api/registerRoute.ts]: they allow type inference for the request and response, based on the validation schemas

File: src/domain/repositories/users.repo.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+export const saveUser = ({ pwd, email }: { pwd: string; email: string }) =>
+    pipe(
+        encryptPassword(pwd),
+        TE.map((pwd_hash: string) => ({
+            pwd_hash,
+            email,
+        })),
+        TE.chain(
+            createEntity<{ email: string; pwd_hash: string }>(ctx, { collection: 'users' })
```
Comment (sidoshi) [inline src/domain/repositories/users.repo.ts]: Why not use the `User` interface from `entities/User`

**Pull request merged**